### PR TITLE
INGK-672 Implement ECAT SDO Network and Servo

### PR DESCRIPTION
**Pull request by MartinALL91**
### Description

Add EtherCAT communication via CoE (SDOs).

### Type of change

- Add scan_slaves function
- Add connect_to_slave function
- Add disconnect_from_slave function
- Add network status listener
- Add subscribe_to_status function
- Add unsubscribe_from_status function
- Add EthercatServo class
- Add functions for reading and writing SDOs
- Add functions to enable monitoring and disturbance

### Tests
- Read from registers
- Write to registers 
- Using ML3 perform a tuning to test that monitoring and disturbance work correctly.

### CODE REVIEW

File: ingenialink/ethercat/network.py [hunk 1]
```diff
@@ -31,19 +73,129 @@ class EthercatNetwork:
...
+        """Disconnects the slave from the network.
+
+        Args:
+            servo: Instance of the servo connected.
+
+        """
+        self.servos.remove(servo)
+        if not self.servos:
+            self.stop_status_listener()
+            self._ecat_master.close()
```
Comment (GullinBustin) [inline ingenialink/ethercat/network.py:147]: If you connect a drive, then disconnect it and try to reconnect a very ugly error appears.
Process finished with exit code -1073741819 (0xC0000005)

File: ingenialink/ethercat/servo.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        self.__slave = slave
+        self.slave_id = slave_id
+        super(EthercatServo, self).__init__(slave.name, dictionary_path, servo_status_listener)
+
+    def _read_raw(
+        self, reg: CanopenRegister, buffer_size: int = 0, complete_access: bool = False
+    ) -> bytes:
+        self._lock.acquire()
+        try:
+            value: bytes = self.__slave.sdo_read(reg.idx, reg.subidx, buffer_size, complete_access)
```
Comment (GullinBustin) [inline ingenialink/ethercat/servo.py:70]: If the wire is disconnected read returns a random value

File: ingenialink/ethercat/servo.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        except Exception as e:
```
Comment (GullinBustin) [inline ingenialink/ethercat/servo.py:71]: Too generic

File: ingenialink/ethercat/servo.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            raise ILIOError(f"Error reading {reg.identifier}. Reason: {e}") from e
+        finally:
+            self._lock.release()
+        return value
+
+    def _write_raw(self, reg: CanopenRegister, data: bytes, complete_access: bool = False) -> None:
+        self._lock.acquire()
+        try:
+            self.__slave.sdo_write(reg.idx, reg.subidx, data, complete_access)
+        except Exception as e:
```
Comment (GullinBustin) [inline ingenialink/ethercat/servo.py:81]: Too generic

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingenialink/ethercat/servo.py (hunk 2 above)
Comment (MartinALL91) [reply ingenialink/ethercat/servo.py:70]: Yes, I saw that. I think that the only thing we can do is to check the slave status and if is not alive, raise an Exception. What do you think?

File: ingenialink/ethercat/network.py (hunk 1 above)
Comment (MartinALL91) [reply ingenialink/ethercat/network.py:147]: Done

File: ingenialink/ethercat/servo.py (hunk 3 above)
Comment (MartinALL91) [reply ingenialink/ethercat/servo.py]: Done

File: ingenialink/ethercat/servo.py (hunk 4 above)
Comment (MartinALL91) [reply ingenialink/ethercat/servo.py]: Done

File: ingenialink/ethercat/servo.py (hunk 2 above)
Comment (MartinALL91) [reply ingenialink/ethercat/servo.py:71]: Solved using the working counter (wkc).

**Review by pzinemanas — APPROVED** [review]

File: ingenialink/ethercat/servo.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    @staticmethod
+    def __monitoring_disturbance_map_can_address(address: int, subnode: int) -> int:
+        """Map CAN register address to IPB register address.
+
+        Args:
+            subnode: Subnode to be targeted.
+            address: Register address to map.
+
+        """
+        return address - (0x2000 + (0x800 * (subnode - 1)))
```
Comment (GullinBustin) [inline ingenialink/ethercat/servo.py:106]: Add 0x2000 and 0x800 as constants

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingenialink/ethercat/servo.py (hunk 5 above)
Comment (MartinALL91) [reply ingenialink/ethercat/servo.py]: Done

**Review by GullinBustin — APPROVED** [review]

**Pull request merged**